storage/innobase/btr/btr0cur.cc

```diff
@@ -669,25 +672,52 @@ btr_cur_will_modify_tree(
 			}
 			/* check delete will cause. (BTR_INTENTION_BOTH
 			or BTR_INTENTION_DELETE) */
-			if (n_recs <= max_nodes_deleted * 2
-			    || page_rec_is_first(rec, page)) {
+			if (n_recs <= max_nodes_deleted * 2) {
 				/* The cursor record can be the left most record
 				in this page. */
 				return true;
 			}
 
-			if (page_has_prev(page)
-			    && page_rec_distance_is_at_most(
-				    page_get_infimum_rec(page), rec,
-				    max_nodes_deleted)) {
-				return true;
-			}
-
-			if (page_has_next(page)
-			    && page_rec_distance_is_at_most(
-				    rec, page_get_supremum_rec(page),
-				    max_nodes_deleted)) {
-				return true;
+			if (page_is_comp(page)) {
+				const rec_t *const infimum
+					= page + PAGE_NEW_INFIMUM;
+				if (page_rec_next_get<true>(page, infimum)
+				    == rec) {
+					return true;
+				}
+				if (!page_has_prev(page)
```

Comment:
Good catch. This is my mistake. Unfortunately this part of the code is not well exercised by our regression test suite. Stress tests should hopefully hit debug assertions that were added to catch potential hangs of the [MDEV-29835](https://jira.mariadb.org/browse/MDEV-29835) type. If not, this could lead to a performance regression due to more frequent escalation to exclusively latching the entire index tree.